File: dmoj/misc/PICK_IT.java
import java.io.*;

public class PICK_IT {
  
  static int [] arr;
  static int [][] cache;
  
  public static int best (int l, int r) {
    if (cache [l][r] == 0 && l + 1 < r) {
      for (int i = l + 1; i < r; i++) {
        cache [l][r] = Math.max (cache [l][r], arr [l] + arr [i] + arr [r] + best (l, i) + best (i, r));//choose number at i, and then find best possible combos for all numbers left of i and right of i
      }
    }
    
    return cache [l][r];
  }
  
  public static void main (String [] t) throws IOException {
    BufferedReader in = new BufferedReader (new InputStreamReader (System.in));
    int N;
    
    while (true) {
      t = in.readLine ().split (" ");
      
      if (t [0].charAt (0) == '0') {
        break;
      }
      
      N = Integer.parseInt (t [0]);
      arr = new int [N];
      cache = new int [N][N];
      
      for (int n = 1; n <= N; n++) {
        arr [n - 1] = Integer.parseInt (t [n]);
      }
      
      System.out.println (best (0, N - 1));
    }
  }
}

File: dmoj/misc/PICK_IT.cpp
#include <stdio.h>
#include <string.h>

#define scan(x) do{while((x=getchar())<'0'); for(x-='0'; '0'<=(_=getchar()); x=(x<<3)+(x<<1)+_-'0');}while(0)
#define max(a,b) ((a)>(b)?(a):(b))
char _;

using namespace std;

int arr [201];
int cache [201][201];

int best (int l, int r) {
	if (!cache [l][r] && l != r) {
		for (int i = l + 1; i < r; i++) {
			cache [l][r] = max (cache [l][r], arr [l] + arr [i] + arr [r] + best (l, i) + best (i, r));
		}
	}
	
	return cache [l][r];
}

int main () {
	int N;
	
	while (true) {
		scan (N);
		
		if (!N) {
			break;
		}
		
		memset (cache, 0, sizeof (cache));
		
		for (int n = 0; n < N; n++) {
			scan (arr [n]);
		}
		
		printf ("%d\n", best (0, N - 1));
	}
}